From ca5c377c10b0e38159c7d1426f01343ca16ae801 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Mon, 21 Jun 2021 07:55:30 -0700
Subject: [PATCH 132/196] Allow mainnet XAddresses to be input into QT address
 fields (#209)

We overlooked that on mainnet underscore characters are part of the
address. The validation of the address fields in QT don't currently
allow underscores. This commit adds them to the list of valid characters
so that we can use the QT interfaces to send and receive mainnet Lotus.
---
 src/qt/bitcoinaddressvalidator.cpp           | 2 +-
 src/qt/test/bitcoinaddressvalidatortests.cpp | 4 ++++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/qt/bitcoinaddressvalidator.cpp b/src/qt/bitcoinaddressvalidator.cpp
index 61542db3f..efe006d29 100644
--- a/src/qt/bitcoinaddressvalidator.cpp
+++ b/src/qt/bitcoinaddressvalidator.cpp
@@ -65,7 +65,7 @@ QValidator::State BitcoinAddressEntryValidator::validate(QString &input,
         int ch = input.at(idx).unicode();
 
         if ((ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'z') ||
-            (ch >= 'A' && ch <= 'Z') || (ch == ':')) {
+            (ch >= 'A' && ch <= 'Z') || (ch == ':') || (ch == '_')) {
             // Alphanumeric and not a 'forbidden' character
             // We also include ':' for cashaddr.
         } else {
diff --git a/src/qt/test/bitcoinaddressvalidatortests.cpp b/src/qt/test/bitcoinaddressvalidatortests.cpp
index eae378d4f..67755bb2b 100644
--- a/src/qt/test/bitcoinaddressvalidatortests.cpp
+++ b/src/qt/test/bitcoinaddressvalidatortests.cpp
@@ -53,4 +53,8 @@ void BitcoinAddressValidatorTests::inputTests() {
     // Only alphanumeric chars are accepted.
     in = "%";
     QVERIFY(v.validate(in, unused) == QValidator::Invalid);
+
+    // valid characters, invalid address
+    in = "lotus_poop";
+    QVERIFY(v.validate(in, unused) == QValidator::Acceptable);
 }
-- 
2.39.5 (Apple Git-154)

